From: srinivas pulukuru <srinivas.pulukuru@ti.com>
Date: Sat, 16 Apr 2011 07:38:52 +0000 (-0500)
Subject: OMAP4 DSS HDMI: Add delay after setting DDC scl clock.
X-Git-Tag: android-2.6.35-2.3-omap4.13.1~29
X-Git-Url: http://git.omapzoom.org/?p=kernel%2Fomap.git;a=commitdiff_plain;h=3b398979a99e227fd7318a74cc0e07174910fb9f

OMAP4 DSS HDMI: Add delay after setting DDC scl clock.

1. Before reading the DDC_STATUS__IN_PROG bit add delay after
   setting the DDC scl clock to allow all devices to reset
   on the DDC I2C bus.
2. Reading the DDC_STATUS__IN_PROG bit immediatley after setting the
   scl clock was seen to cause DDC bus lockup.
3. This was mainly seen on HDMI compliance test equipment. On Blaze
   we were mainly seeing the IN_PROG bit set atleast once before it gets
   cleared. With the delay after scl clock the IN_PROG bit is always 0.

Change-Id: I2ba8f0c569d60606d720c2c222f02cc930d1ec45
Signed-off-by: Srinivas Pulukuru <srinivas.pulukuru@ti.com>
---

diff --git a/arch/arm/plat-omap/hdmi_lib.c b/arch/arm/plat-omap/hdmi_lib.c
index 8d10eb7..17d7745 100644
--- a/arch/arm/plat-omap/hdmi_lib.c
+++ b/arch/arm/plat-omap/hdmi_lib.c
@@ -382,9 +382,15 @@ int hdmi_core_ddc_edid(u8 *pEDID, int ext)
 		/* Clk SCL Devices */
 		REG_FLD_MOD(ins, HDMI_CORE_DDC_CMD, 0xA, 3, 0);
 
+		mdelay(100);
 		/* HDMI_CORE_DDC_STATUS__IN_PROG */
-		while (FLD_GET(hdmi_read_reg(ins, sts), 4, 4) == 1)
-			;
+		while (FLD_GET(hdmi_read_reg(ins, sts), 4, 4) == 1) {
+			printk(KERN_ERR "<%s> ERROR reading DDC Status:"
+					"HDMI_CORE_DDC_STATUS__IN_PROG = "
+					"0x%x\n",
+			       __func__, hdmi_read_reg(ins, sts));
+			mdelay(10);
+		}
 
 		/* Clear FIFO */
 		REG_FLD_MOD(ins, HDMI_CORE_DDC_CMD, 0x9, 3, 0);